src/Orleans.Runtime/Placement/PlacementService.cs

```diff
@@ -179,5 +154,192 @@ public SiloAddress[] AllActiveSilos
+            {
+                var resultTask = completedWorkItem.Result;
+                var messages = completedWorkItem.Messages;
+                if (resultTask.IsCompletedSuccessfully)
+                {
+                    foreach (var message in messages)
+                    {
+                        var result = resultTask.Result;
+                        _placementService.SetMessageTargetPlacement(message.Message, result.Address.Activation, result.Address.Silo, result.IsNewPlacement);
+                        message.Completion.TrySetResult(true);
+                    }
+
+                    messages.Clear();
+                }
+                else
+                {
+                    foreach (var message in messages)
+                    {
+                        message.Completion.TrySetException(resultTask.Exception.OriginalException());
+                    }
+
+                    messages.Clear();
+                }
+            }
+
+            private async Task<(ActivationAddress Address, bool IsNewPlacement)> GetOrPlaceActivationAsync(Message firstMessage)
+            {
+                await Task.Yield();
+                var target = new PlacementTarget(
+                    firstMessage.TargetGrain,
+                    firstMessage.RequestContextData,
+                    firstMessage.InterfaceType,
+                    firstMessage.InterfaceVersion);
+
+                var targetGrain = target.GrainIdentity;
+                var result = await _placementService._grainLocator.Lookup(targetGrain);
+                if (result is not null)
+                {
+                    return (result, false);
+                }
+
+                var strategy = _placementService._strategyResolver.GetPlacementStrategy(target.GrainIdentity.Type);
+                var director = _placementService._directorResolver.GetPlacementDirector(strategy);
+                var siloAddress = await director.OnAddActivation(strategy, target, _placementService);
+                
+                // Give the grain locator one last chance to tell us that the grain has already been placed
+                if (_placementService._grainLocator.TryCachedLookup(targetGrain, out result))
+                {
+                    return (result, false);
+                }
+
+                ActivationId activationId;
+                if (strategy.IsDeterministicActivationId)
+                {
+                    // Use the grain id as the activation id.
+                    activationId = ActivationId.GetDeterministic(target.GrainIdentity);
+                }
+                else
+                {
+                    activationId = ActivationId.NewId();
+                }
+
+                result = ActivationAddress.GetAddress(siloAddress, targetGrain, activationId);
+                _placementService._grainLocator.InvalidateCache(targetGrain);
+                _placementService._grainLocator.CachePlacementDecision(result);
```

Comment:
If the grain doesn't end up activated on the target silo, you will have a wrong entry in the cache for a long time. I think we have the same issue if a grain is deactivated after placement, but I feel here we are making things worse.

I believe right now we only invalidate cache entries form the `LocalGrainDirectory`. We should change that in this PR.

Reply:
The entry will be invalidated by return from the call made to the remote silo (forwarded to the right silo, hopefully, or rejected back to the caller)

Reply:
Addressed in latest